olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,191 @@
+from olive.common.utils import get_credentials, hash_dict
+from olive.model.config.model_config import ModelConfig
+
+logger = logging.getLogger(__name__)
+
+
+class CloudCacheConfig(ConfigBase):
+    enable_cloud_cache: bool = True
+    account_url: str = "https://olivepublicmodels.blob.core.windows.net"
+    contaier_name: str = "olivecachemodels"
+    upload_to_cloud: bool = True
+    input_model_config: ModelConfig = None
+
+
+class CloudCacheHelper:
+    def __init__(
+        self,
+        cache_dir: str,
+        account_url: str,
+        container_name: str,
+        hf_cache_path: str,
+        input_model_config: ModelConfig = None,
+    ):
+        try:
+            from azure.storage.blob import ContainerClient
+        except ImportError as exc:
+            raise ImportError(
+                "Please install azure-storage-blob and azure-identity to use the cloud model cache feature."
+            ) from exc
+        credential = get_credentials()
+        self.container_name = container_name
+        self.container_client = ContainerClient(
+            account_url=account_url, container_name=self.container_name, credential=credential
+        )
+
+        self.cache_dir = cache_dir
+        self.output_model_path = Path(cache_dir) / "model"
+        self.input_model_config = input_model_config
+        self.hf_cache_path = hf_cache_path
+
+    def update_model_config(self, cloud_model_path, model_config, input_model_hash):
+        logger.info("Updating model config with cloud model path: %s", cloud_model_path)
+        model_directory_prefix = f"{input_model_hash}/model"
+        blob_list = self.container_client.list_blobs(name_starts_with=model_directory_prefix)
+
+        for blob in blob_list:
+            blob_client = self.container_client.get_blob_client(blob)
+            local_file_path = self.output_model_path / blob.name[len(model_directory_prefix) + 1 :]
+            logger.info("Downloading %s to %s", blob.name, local_file_path)
+            local_file_path.parent.mkdir(parents=True, exist_ok=True)
+            with open(local_file_path, "wb") as download_file:
+                download_stream = blob_client.download_blob()
+                download_file.write(download_stream.readall())
+
+        model_config.config["model_path"] = str(self.output_model_path / cloud_model_path)
+
+        return model_config
+
+    def get_hash_key(self, model_config: ModelConfig, pass_search_point: Dict[str, Any], input_model_hash: str):
+        hf_hub_model_commit_id = None
+        model_config_copy = deepcopy(model_config)
+        if input_model_hash is None:
+            from huggingface_hub import repo_info
+
+            if model_config.has_hf_config():
+                hf_hub_model_commit_id = repo_info(model_config.get_hf_model_name()).sha
+        else:
+            if model_config_copy.config.get("model_path") is not None:
+                model_config_copy.config["model_path"] = None
```

Comment:
The assignment of `model_path` should be unconditional otherwise you will end up with two different hashes for same model - one with `model_path: None` and another with no `model_path` which but both really mean the same for this practical use case.

Ideally, I would just replace that whole else part with `model_config_copy.config.pop("model_path", None)` so the `model_path` is never taken into account for computing the hash.

Reply:
This sounds good to me. Will update this.